### PR TITLE
add iir{notch,comb,peak} design functions

### PR DESCRIPTION
**Pull request by ev-br**
cross-ref #7403

Port IIR filter design functions:

- [x] iircomb
- [x] iirnotch
- [x] iirpeak

<s>This PR is on top of gh-7632<s> rebased.

### CODE REVIEW

**Comment by ev-br** [comment]
rebased on gh-7632

**Comment by ev-br** [comment]
Rebases on main, now that gh-7632 is in. Ready for review from my side.

File: cupyx/scipy/signal/_iir_filter_design.py [hunk 1]
```diff
@@ -650,6 +650,303 @@ def iirdesign(wp, ws, gpass, gstop, analog=False, ftype='ellip', output='ba',
...
+    # formulas 11.3.7 (p.575) and 11.3.21 (p.579)
+    # from reference [1]
+    if ftype == "notch":
+        b = gain * cupy.r_[1.0, -2.0 * cupy.cos(w0), 1.0]
+    else:
+        b = (1.0 - gain) * cupy.r_[1.0, 0.0, -1.0]
+
+    a = cupy.r_[1.0, -2.0 * gain * cupy.cos(w0), 2.0 * gain - 1.0]
+
+    return b, a
```
Comment (takagi) [inline cupyx/scipy/signal/_iir_filter_design.py:947]: How about computing them on CPU and `cupy.asarray` just before return as the inputs are Python scalars and the data size is small.

File: cupyx/scipy/signal/_iir_filter_design.py [hunk 2]
```diff
@@ -650,6 +650,303 @@ def iirdesign(wp, ws, gpass, gstop, analog=False, ftype='ellip', output='ba',
...
+    w_delta = w0 / Q
+
+    # Define base gain values depending on notch or peak filter
+    # Compute -3dB attenuation
+    # Eqs. 11.4.1 and 11.4.2 (p. 582) from reference [1]
+    if ftype == 'notch':
+        G0, G = 1, 0
+    elif ftype == 'peak':
+        G0, G = 0, 1
+    GB = 1 / cupy.sqrt(2)
```
Comment (takagi) [inline cupyx/scipy/signal/_iir_filter_design.py:752]: Computing on CPU, then setting values to `a` and `b` on GPU would be okay.

File: cupyx/scipy/signal/_iir_filter_design.py [hunk 3]
```diff
@@ -650,6 +650,303 @@ def iirdesign(wp, ws, gpass, gstop, analog=False, ftype='ellip', output='ba',
...
+
+    # Compute beta
+    # Eq. 11.5.3 (p. 591) from reference [1]
+    beta = cupy.sqrt((GB**2 - G0**2) / (G**2 - GB**2)) * \
+        cupy.tan(N * w_delta / 4)
```
Comment (takagi) [inline cupyx/scipy/signal/_iir_filter_design.py:757]: ditto (Computing on CPU)

**Comment by ev-br** [comment]
Indeed, it only appears in scipy 1.9.x. Will update the PR.

**Comment by ev-br** [comment]
Pushed the updated PR: move scalar computations to CPU + guard a test for `iirnotch(..., pass_zero=...)` to only run with scipy >= 1.9.0.

**Review by takagi — APPROVED** [review]
LGTM!

**Comment by takagi** [comment]
Would you check `cupy.linux.cuda112` result, which uses scipy==1.7?

**Comment by ev-br** [comment]
Indeed, the failing test was checking a bug which was only fixed in scipy 1.9.x (https://github.com/scipy/scipy/pull/16145). Added a commit to skip it.

**Review by takagi — APPROVED** [review]

**Comment by takagi** [comment]
Thanks! I'll merge this after CI finishes.

**Comment by takagi** [comment]
Thanks, @ev-br!

**Pull request merged**